dynamic_programming/levenshtein_distance.nim

```diff
@@ -0,0 +1,68 @@
+## Levenshtein distance
+
+import tables
+
+
+func toKey(indA, indB: int): (Natural, Natural) =
+  return (indA.Natural, indB.Natural)
```

Comment:
Dot call is a personal preference which I tend to use mostly for subtypes, parsing internally the same as a literal modifier `N'uint`.

Unfortunately, converters are only available at the top level. If they were available at a block scope, the top of the `computeLevenshteinDistanceMatrix` would totally be the place to add it, but alas. 

However, I have another hackish idea. Hackish, but not compromising any guarantees the code had previously, even when compiled without the assertion. No need for `toKey`, `initSubsolutions` or any converters any more.

```nim
func computeLevenshteinDistanceMatrix(a, b: string): Table[(Natural, Natural),
                                                           Natural] =
  func `+`(a: Natural; b: int): Natural = assert b>=0; system.`+`(a, Natural(b))
  for indA in Natural(0)..a.len:
    result[(indA, Natural(0))] = indA

  for indB in Natural(1)..b.len:
    result[(Natural(0), indB)] = indB

  for indA in Natural(0)..<a.len:
    for indB in Natural(0)..<b.len:
      let substitutionCost = (if a[indA] == b[indB]: 0 else: 1)
      result[(indA + 1, indB + 1)] = [
           result[(indA, indB + 1)] + 1,
           result[(indA + 1, indB)] + 1,
           result[(indA, indB)] + substitutionCost].min()
```

Reply:
Local operator overloading, I see. I prefer the toKey solution as it looks more readable.

Reply:
Honestly, `toKey` here is just noise. Key here is just a tuple, not even a named type.

We know that the indexes are natural, and adding a positive literal number to it can't result in a negative (as we usually pretend with integers). So, all we do here is making the compiler happy. We can even make `b` in the new `+` static to ensure that no one accidentally puts a negative there _(added in an edit_).

So, you just pretend the `+` is not there and then code reads just fine, to me it looks like an improvement.

PS: to be frank, I'm not even sure why does the compiler treat the second field in a `(indA, indB + 1)` as `int`, while `result[...] + 1` inside the array in the final statement are fine to be assigned to a Natural key of `result`. In both instances we have a `Natural(N) + 1`. 🤷

Reply:
> you just pretend the + is not there

That is exactly the problem I have. I want people to be able to understand each line and an operator overloading looks like something superfluous for this algorithm.
You are right though, `Natural` and `toKey` bring much noise.

In some sense, we have to tell the compiler that there will be no wrapping since everything is bounded by `max(lenA, lenB)`.
Another option would be to fall back on using `int` instead. After all, what is the type of `len(A)`? `int` or `Natural`?

Reply:
> After all, what is the type of `len(A)`? `int` or `Natural`?

That's one of the sources of the issue. `len` is `int` nominally, but `Natural` de facto[^r]. Looks like an oversight, but it's a choice made to smoothen out the API of standard library (which probably started to form before range types, but I'm not sure).

>  I want people to be able to understand each line

There's an easy solution ;) Just comment that `+` func.

[^r]:  Compare to indexes in Rust which are always unsigned `usize`.

Reply:
Ok for that `+` overload but it should be commented heavily.
-> A beginner does not necessary know about that `+` is an overload, that the strokes are not significant.
-> Why is this overload needed here?
-> Why there is no need to overload the other pair of arguments (a: static int, b: Natural). Is it due to the algorithm only, or is it Nim specific?
-> Why is the `static` keyword needed?
4 questions, so at least four lines of comments for one line.

Reply:
https://github.com/nim-lang/Nim/issues/16839 and https://status-im.github.io/nim-style-guide/language.range.html more generally should be watched closely in any case.

Reply:
I think you're a bit overthinking it. Firstly, some of the questions are better answered in a commit message. Secondly, it's not always necessary to explain everything to the lowest level of comprehension, sometimes it's fine to leave just information to infer the meaning or explore when in doubt. Or, if you treat the code as educational, it's a good place to explain the technique and its merits.

```nim
  func `+`(a: Natural; b: static int): Natural =
    ## An ad-hoc overload to ensure the compiler that in all cases below a sum
    ## of a Natural and the `1` literal (thus `static`) is always Natural.
    assert b>=0; system.`+`(a, Natural(b))
```

Let's also not forget, it's an internal function where the input can't ever be wrong unless there's a bug up the call stack. We can leave as many comments here as necessary, as (if it was a production code) they are only there for the next guy to refactor this piece.

Below is how I think the code can look if we get back to the previous implementation and just clean it a bit (the amount of changes is minuscule). Does it look right and correct from the first glance? Yes. Does it require additional commentaries explaining code org. decisions? Probably no. Does it actually show what's happening better than a version with overloads? I don't think so. It trades a self-commenting `DistKey` type for a level of indirection and worse locality (and in a longer and more complex function it _would_ actually matter). The whole purpose of the type here is the same: please the compiler into using a properly restricting type without manual conversion. It does just that (and has a name). What you also gain is less noisy func body. Does this justify the type's existence? I don't know.

```nim
type DistKey = (Natural, Natural)

func toKey(nA, nB: int): DistKey = (Natural(nA), Natural(nB))

func computeLevenshteinDistanceMatrix(a, b: string): Table[DistKey, Natural] =
  for indA in 0..a.len:
    result[toKey(indA, 0)] = indA

  for indB in 1..b.len:
    result[toKey(0, indB)] = indB

  for indA in 0..<a.len:
    for indB in 0..<b.len:
      let substitutionCost = if a[indA] == b[indB]: 0 else: 1
      result[toKey(indA + 1, indB + 1)] = [
           result[toKey(indA, indB + 1)] + 1,
           result[toKey(indA + 1, indB)] + 1,
           result[toKey(indA, indB)] + substitutionCost].min()
```

Reply:
I really like the idea of introducing the _KeyType_ (I was thinking that I will be told that it is an overkill) (cf. ec122d0c3a12b49f56754a49f3f4e4c701b2c184). Although the notation `x.Natural` feels awkward it is one character shorten (than `Natural(x)`) and it dies not introduce any parentheses.

I decided to leave the `initSubsolutions` function (in order to make `computeLevenshteinDistanceMatrix` shorter). I am hesitant to introduce a procedure like `proc fillLevenshteinDistanceMatrix(someTable: var Table[Key, Natural]` in order to have _same level of abstraction_ in `computeLevenshteinDistanceMatrix`.

@dlesnoff  If you think that there is noting more to change here, please mark it as _resolved_.